planner/core/stats.go

```diff
@@ -437,26 +448,22 @@ func (ds *DataSource) accessPathsForConds(conditions []expression.Expression, us
 }
 
 // buildIndexMergePartialPath chooses the best index path from all possible paths.
-// Now we just choose the index with most columns.
-// We should improve this strategy, because it is not always better to choose index
-// with most columns, e.g, filter is c > 1 and the input indexes are c and c_d_e,
-// the former one is enough, and it is less expensive in execution compared with the latter one.
-// TODO: improve strategy of the partial path selection
-func (ds *DataSource) buildIndexMergePartialPath(indexAccessPaths []*util.AccessPath) *util.AccessPath {
+// Now we choose the index with minimal estimate row count.
+func (ds *DataSource) buildIndexMergePartialPath(indexAccessPaths []*util.AccessPath) (*util.AccessPath, error) {
 	if len(indexAccessPaths) == 1 {
-		return indexAccessPaths[0]
+		return indexAccessPaths[0], nil
 	}
 
-	maxColsIndex := 0
-	maxCols := len(indexAccessPaths[0].IdxCols)
-	for i := 1; i < len(indexAccessPaths); i++ {
-		current := len(indexAccessPaths[i].IdxCols)
-		if current > maxCols {
-			maxColsIndex = i
-			maxCols = current
+	minEstRowIndex := 0
+	minEstRow := math.MaxFloat64
+	for i := 0; i < len(indexAccessPaths); i++ {
+		rc := indexAccessPaths[i].CountAfterAccess
+		if rc < minEstRow {
+			minEstRowIndex = i
+			minEstRow = rc
```

Comment:
Shall we cherry-pick the full changes of https://github.com/pingcap/tidb/pull/23141 to this PR? otherwise the code / behavior may be different on 4.0 / master then.

Reply:
Ok, but the explain test is wrong, do you have any idea?